_episodes_rmd/01-raster-structure.Rmd

```diff
@@ -1,406 +1,498 @@
+
+By default the shape of a raster is always rectangular. So if we have  a dataset
+that has a shape that isn't rectangular, some pixels at the edge of the raster
+will have `NoDataValue`s. This often happens when the data were collected by an
+airplane which only flew over some part of a defined region.
+
+In the image below, the pixels that are black have `NoDataValue`s. The camera
+did not collect data in these areas.
+
+```{r demonstrate-no-data-black-ggplot, echo=FALSE }
+# no data demonstration code - not being taught
+# Use stack function to read in all bands
+RGB_stack <-
+  stack("data/NEON-DS-Airborne-Remote-Sensing/HARV/RGB_Imagery/HARV_RGB_Ortho.tif")
+
+# aggregate eclls from 0.25m to 2m for plotting to speed up the lesson and 
+# save memory
+RGB_2m <- raster::aggregate(RGB_stack, fact = 8, fun = median)
+# fix data values back to integer datatype
+values(RGB_2m) <- as.integer(round(values(RGB_2m)))
+
+# convert to a df for plotting using raster's built-in methods
+RGB_2m_df  <- raster::as.data.frame(RGB_2m, xy = TRUE)
+# make colnames easy to ref
+names(RGB_2m_df) <- c('x', 'y', 'red', 'green', 'blue')
+
+ggplot() +
+ geom_raster(data = RGB_2m_df , aes(x = x, y = y, fill = red),
+             show.legend = FALSE) +
+  scale_fill_gradient(low = 'black', high = "red") +
+  ggtitle("Orthographic Imagery", subtitle = 'Red Band') +
+  coord_equal()
+# more memory saving
+#rm(RGB_df)
+```
+
+In the next image, the black edges have been assigned `NoDataValue`. `R` doesn't
+render pixels that contain a specified `NoDataValue`. `R` assigns missing data
+with the `NoDataValue` as `NA`.
+
+The difference here shows up as ragged edges on the plot, rather than black
+spaces where there is no data.
+
+```{r demonstrate-no-data-ggplot, echo=FALSE}
+RGB_2m_df_nd <- RGB_2m_df
+# convert the three rgb values to hex codes
+RGB_2m_df_nd$hex <- rgb(RGB_2m_df_nd$red,
+                        RGB_2m_df_nd$green,
+                        RGB_2m_df_nd$blue, maxColorValue = 255)
+# set black hex code to NA
+RGB_2m_df_nd$hex[RGB_2m_df_nd$hex == '#000000'] <- NA_character_ 
+
+ggplot() +
+  geom_raster(data = RGB_2m_df_nd, aes(x = x, y = y, fill = hex)) +
+  scale_fill_identity() +
+  ggtitle("Orthographic Imagery", subtitle = "All bands") +
+  coord_equal()
+
+```
+
+If your raster already has NA values set correctly but you aren't sure where they are, you can use `ggspatial` to deliberately plot them in a particular colour. This can be useful when checking a dataset's coverage. For instance, sometimes data can be missing where a sensor could not 'see' its target data, and you may wish to locate that missing data and fill it in.
+
+To highlight NA values in gpplot, alter the `scale_fill_*()` layer to contain a colour instruction for na values, like `scale_fill_viridis_c(na.value = 'deeppink')`
+
+```{r 'napink', eval = FALSE}
+# function to replace 0 with NA where all three values are 0 only
+RGB_2m_nas <- calc(RGB_2m, 

```

Comment:
This is pretty advanced. The learners haven't seen function definition syntax yet. This also introduces three new functions: `calc()`, `rowSums()`, and `rep()`.